planner/cascades/implementation_rules.go

```diff
@@ -145,23 +149,71 @@ func (r *ImplTiKVSingleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *pr
 	return impl.NewTableReaderImpl(reader, sg.Source.TblColHists), nil
 }
 
+// ImplTiKVDoubleReadGather implements TiKVDoubleGather as PhysicalIndexLookUpReader.
+type ImplTiKVDoubleReadGather struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTiKVDoubleReadGather) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	if prop.IsEmpty() {
+		return true
+	}
+	reader := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	if all, _ := prop.AllSameOrder(); all {
+		for i, col := range reader.IndexCols {
+			if col.Equal(nil, prop.Items[0].Col) {
+				return plannercore.MatchIndicesProp(reader.IndexCols[i:], reader.IndexColLens[i:], prop.Items)
+			}
+		}
+	}
+	return false
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplTiKVDoubleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	logicProp := expr.Group.Prop
+	sg := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	var reader plannercore.PhysicalPlan
+	reader = sg.GetPhysicalIndexLookUpReader(logicProp.Schema, logicProp.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt), reqProp, reqProp)
+	if sg.Source.HandleCol.ID == model.ExtraHandleID && !reqProp.IsEmpty() {
```

Comment:
What if the schema already contains the `ExtraHanldeColumn`?

Reply:
Do you mean selecting `tidb_rowid` in the SQL?

Reply:
Yes

Reply:
It doesn't matter. Here double read will generate a projection for the original schema. I prefer to add a `TODO` here because there are some possible situations we should take into consideration. In the future, we can handle the enhancement.